File: src/main/kotlin/_2_null_saftey/exercise/null_not_allowed/NonNullable.java
package _2_null_saftey.exercise.null_not_allowed;


public class NonNullable {
    public void sendMessageToClient(
            ClientJ client,
            String message,
            MailerJ mailer
    ) {
        if (client == null) {
            throw new NullPointerException("client must be non null");
        }
        if (message == null) {
            throw new NullPointerException("message must be non null");
        }

        PersonalInfoJ personalInfo = client.getPersonalInfo();
        String email = personalInfo.getEmail();

        mailer.sendMessage(email, message);
    }


    public static void main(String[] args) {
        NonNullable nullableTypes = new NonNullable();

        ClientJ mail = new ClientJ(new PersonalInfoJ("Mail"));
        nullableTypes.sendMessageToClient(mail, "Java hello kata", (email, message) -> System.out.println(message));

        //Does compile, fails at runtime
        ClientJ failed = new ClientJ(null);
        nullableTypes.sendMessageToClient(null, "Java hello kata", (email, message) -> System.out.println(message));
    }
}

class ClientJ {
    private final PersonalInfoJ personalInfo;

    public ClientJ(PersonalInfoJ personalInfo) {
        if(personalInfo == null){
            throw new NullPointerException("personalInfo must be non null");
        }

        this.personalInfo = personalInfo;
    }

    public PersonalInfoJ getPersonalInfo() {
        return personalInfo;
    }
};

class PersonalInfoJ {
    private final String email;

    public PersonalInfoJ(String email) {
        if(email == null){
            throw new NullPointerException("personalInfo must be non null");
        }

        this.email = email;
    }

    public String getEmail() {
        return email;
    }
};

interface MailerJ {
    void sendMessage(String email, String message);
}

File: src/main/kotlin/_2_null_saftey/exercise/null_not_allowed/NonNullable.kt
package _2_null_saftey.exercise.null_not_allowed

/**
 * Java programmer that wrote null_allowed had revelation about why nulls are bad and decided to remove them from his codebase.
 *
 * Disallow nulls in bellow code and compare it to NonNullable java code
 */
fun sendMessageToClient(
    client: Client?, message: String?, mailer: Mailer
) {
    val personalInfo = client?.personalInfo
    val email = personalInfo?.email

    if (email == null || message == null) return

    mailer.sendMessage(email, message)
}

class Client(val personalInfo: PersonalInfo?)
class PersonalInfo(val email: String?)
interface Mailer {
    fun sendMessage(email: String, message: String)
}

fun main() {
    val personalInfo = PersonalInfo(null);  //Should fail at compilation
    val client = Client(null)          //Should fail at compilation
    val anonymousMailer = object : Mailer {
        override fun sendMessage(email: String, message: String) = println("Sending message to $email")
    }
    sendMessageToClient(null, "hey kata", anonymousMailer)  //Should fail at compilation
    sendMessageToClient(client, null, anonymousMailer)            //Should fail at compilation



}
